cr-usecases/src/main/java/com/thefuntasty/mvvm/crusecases/Result.kt

```diff
@@ -0,0 +1,270 @@
+package com.thefuntasty.mvvm.crusecases
+
+import kotlinx.coroutines.CancellationException
+
+/**
+ * A discriminated union that encapsulates successful outcome either with [Success] or with [Error]
+ *
+ * This class is adapted to kotlin coroutines so if [Error] contains [CancellationException] then this error is rethrown.
+ */
+sealed class Result<out VALUE : Any?> {
+    override fun toString(): String {
+        return when (this) {
+            is Success -> "Success($value)"
+            is Error -> "Error(${error})"
+        }
+    }
+
+    operator fun component1(): VALUE? = if (this is Success) value else null
+
+    operator fun component2(): Throwable? = if (this is Error) error else null
+}
+
+/**
+ * Represents a successful [Result], containing a [value].
+ */
+class Success<VALUE : Any?>(val value: VALUE) : Result<VALUE>() {
+
+    override fun equals(other: Any?): Boolean {
+        if (this === other) {
+            return true
+        }
+        if (other == null || this::class != other::class) {
+            return false
+        }
+
+        other as Success<*>
+
+        if (value != other.value) {
+            return false
+        }
+
+        return true
+    }
+
+    override fun hashCode() = value.hashCode()
+}
+
+/**
+ * Represents a failed [Result], containing an [error].
+ */
+class Error(val error: Throwable) : Result<Nothing>() {
+
+    override fun equals(other: Any?): Boolean {
+        if (this === other) {
+            return true
+        }
+        if (other == null || this::class != other::class) {
+            return false
+        }
+
+        other as Error
+
+        if (error != other.error) {
+            return false
+        }
+
+        return true
+    }
+
+    override fun hashCode() = error.hashCode()
+}
+
+// Functions
+
+/**
+ * Calls the specified function [block] and returns [Success] with encapsulated result if invocation was successful,
+ * catching and returning [Error] with any thrown exception except [CancellationException].
+ */
+inline fun <VALUE> tryCatch(block: () -> VALUE): Result<VALUE> {
+    return try {
+        Success(block())
+    } catch (error: Throwable) {
+        if (error is CancellationException) {
```

Comment:
<table data-meta="generated_by_danger-pr">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/thefuntasty/mvvm-android/blob/01753e4e83475541b61b01cfce8108bc8990ca63/cr-usecases/src/main/java/com/thefuntasty/mvvm/crusecases/Result.kt#L83"></span>Instead of checking for a general exception type and checking for a specific exception type use multiple catch blocks.</td>
    </tr>
  </tbody>
</table>